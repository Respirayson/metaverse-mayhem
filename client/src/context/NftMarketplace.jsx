import React, { useEffect, useState, createContext } from 'react';
import { ethers } from 'ethers';

import {
  marketplaceABI,
  marketplaceAddress,
  contractAddress,
} from '../utils/constants';
import { getEthereumContract } from '../utils/connect';

// Create a context for the NFT marketplace
export const NftMarketplaceContext = createContext();

const { ethereum } = window;

/**
 * Component representing the NFT marketplace provider
 * @param {Object} props - The component props
 * @param {ReactNode} props.children - The children to be rendered inside the provider
 * @returns {JSX.Element} - The JSX element
 */
export function NftMarketplaceProvider({ children }) {
  const [currentAccount, setCurrentAccount] = useState(null);

  /**
   * Check if the wallet is connected and set the current account
   */
  const checkIfWalletIsConnected = async () => {
    try {
      if (!ethereum) {
        return;
      }

      const accounts = await ethereum.request({ method: 'eth_accounts' });

      if (accounts.length) {
        const account = accounts[0];
        setCurrentAccount(account);
      } else {
        // No account connected
      }
    } catch (err) {
      console.log(err);
    }
  };

  /**
   * Add a listing for an NFT
   * @param {Object} collectible - The NFT collectible object
   * @param {number} price - The price of the NFT
   */
  const addListing = async (collectible, price) => {
    if (ethereum) {
      const contract = getEthereumContract(marketplaceAddress, marketplaceABI);
      const parsedPrice = ethers.utils.parseEther(price.toString());
      const transaction = await contract.listItem(
        contractAddress,
        collectible.tokenId,
        parsedPrice,
      );
      await transaction.wait();
      console.log(
        `Successfully listed NFT with id ${collectible.tokenId} for ${price} ETH`,
      );

      await fetch('https://metaverse-mayhem.onrender.com/api/v1/marketplace', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          tokenId: collectible.tokenId,
          price,
          seller: currentAccount,
          cardId: collectible.card.id,
        }),
      });
      console.log('Successfully listed NFT');
    } else {
      console.log('Ethereum is not present');
    }
  };

  /**
   * Buy an NFT
   * @param {string} listingId - The ID of the NFT listing
   * @param {string} tokenId - The ID of the NFT token
   */
  const buyItem = async (listingId, tokenId) => {
    if (ethereum) {
      const contract = getEthereumContract(marketplaceAddress, marketplaceABI);
      const listing = await contract.getListing(contractAddress, tokenId);
      const transaction = await contract.buyItem(contractAddress, tokenId, {
        value: listing.price,
      });
      await transaction.wait();
      console.log(`Successfully bought NFT with id ${tokenId}`);
      await fetch(
        `https://metaverse-mayhem.onrender.com/api/v1/marketplace/${listingId}`,
        {
          method: 'DELETE',
        },
      );
      console.log('Successfully deleted listing');
    }
  };

  /**
   * Get the proceeds from the NFT marketplace
   * @returns {number} - The proceeds in ETH
   */
  const getProceeds = async () => {
    if (ethereum) {
      const contract = getEthereumContract(marketplaceAddress, marketplaceABI);
      const transaction = await contract.getProceeds(currentAccount);
      return ethers.utils.formatEther(transaction);
    }
    return 0;
  };

  /**
   * Withdraw the proceeds from the NFT marketplace
   */
  const withdrawProceeds = async () => {
    if (ethereum) {
      const contract = getEthereumContract(marketplaceAddress, marketplaceABI);
      const transaction = await contract.withdrawProceeds();
      await transaction.wait();
      console.log('Successfully withdrew proceeds');
    }
  };

  // Check if the wallet is connected on component mount
  useEffect(() => {
    checkIfWalletIsConnected();
  }, []);

  return (
    <NftMarketplaceContext.Provider
      value={{
        addListing,
        buyItem,
        currentAccount,
        getProceeds,
        withdrawProceeds,
      }}
    >
      {children}
    </NftMarketplaceContext.Provider>
  );
}
