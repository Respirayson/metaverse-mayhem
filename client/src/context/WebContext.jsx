import React, { useState, useEffect, createContext } from 'react';

export const WebContext = createContext();

export function WebProvider({ children }) {
  const [showAlert, setShowAlert] = useState(false);
  const [alertMessage, setAlertMessage] = useState('');
  const [success, setSuccess] = useState(false);

  useEffect(() => {
    if (showAlert) {
      const timer = setTimeout(() => {
        setShowAlert(false);
        setAlertMessage("");
        setSuccess(false);
      }, [3500]);

      return () => clearTimeout(timer);
    }
  }, [showAlert]);

  return (
    <WebContext.Provider
      value={{
        showAlert,
        setShowAlert,
        alertMessage,
        setAlertMessage,
        success,
        setSuccess,
      }}
    >
      {children}
    </WebContext.Provider>
  );
}
