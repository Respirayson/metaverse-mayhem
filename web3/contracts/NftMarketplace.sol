// SPDX-License-Identifier: MIT
pragma solidity ^0.8.4;

import "@openzeppelin/contracts/token/ERC721/IERC721.sol";
import "@openzeppelin/contracts/security/ReentrancyGuard.sol";

/**
 * @title MMT Marketplace Contract
 * @notice A decentralized marketplace for buying and selling Metaverse Mayhem Cards.
 */
contract NftMarketplace is ReentrancyGuard {
    struct Listing {
        uint256 price;
        address seller;
    }

    // Events
    event ItemListed(
        address indexed seller,
        address indexed nftAddress,
        uint256 indexed tokenId,
        uint256 price
    );
    event ItemCanceled(
        address indexed seller,
        address indexed nftAddress,
        uint256 indexed tokenId
    );
    event ItemBought(
        address indexed buyer,
        address indexed nftAddress,
        uint256 indexed tokenId,
        uint256 price
    );

    mapping(address => mapping(uint256 => Listing)) private s_listings;
    mapping(address => uint256) private s_proceeds;

    // Modifiers

    /**
     * @dev Modifier to check if an NFT is not listed for sale.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     */
    modifier notListed(address nftAddress, uint256 tokenId) {
        Listing memory listing = s_listings[nftAddress][tokenId];
        if (listing.price > 0) {
            revert("AlreadyListed");
        }
        _;
    }

    /**
     * @dev Modifier to check if an NFT is listed for sale.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     */
    modifier isListed(address nftAddress, uint256 tokenId) {
        Listing memory listing = s_listings[nftAddress][tokenId];
        if (listing.price <= 0) {
            revert("NotListed");
        }
        _;
    }

    /**
     * @dev Modifier to check if the caller is the owner of an NFT.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     * @param spender Address of the caller.
     */
    modifier isOwner(
        address nftAddress,
        uint256 tokenId,
        address spender
    ) {
        IERC721 nft = IERC721(nftAddress);
        address owner = nft.ownerOf(tokenId);
        if (spender != owner) {
            revert("NotOwner");
        }
        _;
    }

    // Main Functions

    /**
     * @notice Lists an NFT for sale on the marketplace.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     * @param price Sale price for the NFT.
     */
    function listItem(
        address nftAddress,
        uint256 tokenId,
        uint256 price
    )
        external
        notListed(nftAddress, tokenId)
        isOwner(nftAddress, tokenId, msg.sender)
    {
        if (price <= 0) {
            revert("PriceMustBeAboveZero");
        }
        IERC721 nft = IERC721(nftAddress);
        if (nft.getApproved(tokenId) != address(this)) {
            revert("NotApprovedForMarketplace");
        }
        s_listings[nftAddress][tokenId] = Listing(price, msg.sender);
        emit ItemListed(msg.sender, nftAddress, tokenId, price);
    }

    /**
     * @notice Cancels a listing for an NFT.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     */
    function cancelListing(
        address nftAddress,
        uint256 tokenId
    )
        external
        isOwner(nftAddress, tokenId, msg.sender)
        isListed(nftAddress, tokenId)
    {
        delete (s_listings[nftAddress][tokenId]);
        emit ItemCanceled(msg.sender, nftAddress, tokenId);
    }

    /**
     * @notice Buys a listed NFT.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     */
    function buyItem(
        address nftAddress,
        uint256 tokenId
    ) external payable isListed(nftAddress, tokenId) nonReentrant {
        Listing memory listedItem = s_listings[nftAddress][tokenId];
        if (msg.value < listedItem.price) {
            revert("PriceNotMet");
        }
        s_proceeds[listedItem.seller] += msg.value;
        delete (s_listings[nftAddress][tokenId]);
        IERC721(nftAddress).safeTransferFrom(
            listedItem.seller,
            msg.sender,
            tokenId
        );
        emit ItemBought(msg.sender, nftAddress, tokenId, listedItem.price);
    }

    /**
     * @notice Updates the price of a listed NFT.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     * @param newPrice New price for the NFT.
     */
    function updateListing(
        address nftAddress,
        uint256 tokenId,
        uint256 newPrice
    )
        external
        isListed(nftAddress, tokenId)
        nonReentrant
        isOwner(nftAddress, tokenId, msg.sender)
    {
        if (newPrice <= 0) {
            revert("PriceMustBeAboveZero");
        }
        s_listings[nftAddress][tokenId].price = newPrice;
        emit ItemListed(msg.sender, nftAddress, tokenId, newPrice);
    }

    /**
     * @notice Withdraws proceeds from sales.
     */
    function withdrawProceeds() external {
        uint256 proceeds = s_proceeds[msg.sender];
        if (proceeds <= 0) {
            revert("NoProceeds");
        }
        s_proceeds[msg.sender] = 0;
        (bool success, ) = payable(msg.sender).call{value: proceeds}("");
        require(success, "Transfer failed");
    }

    // Getter Functions

    /**
     * @notice Gets the listing details of an NFT.
     * @param nftAddress Address of the NFT contract.
     * @param tokenId Token ID of the NFT.
     * @return Listing details of the NFT.
     */
    function getListing(
        address nftAddress,
        uint256 tokenId
    ) external view returns (Listing memory) {
        return s_listings[nftAddress][tokenId];
    }

    /**
     * @notice Gets the total proceeds of a seller.
     * @param seller Address of the seller.
     * @return Total proceeds of the seller.
     */
    function getProceeds(address seller) external view returns (uint256) {
        return s_proceeds[seller];
    }
}
