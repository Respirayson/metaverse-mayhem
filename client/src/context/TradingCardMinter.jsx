import React, {
  useEffect,
  useState,
  useRef,
  createContext,
  useContext,
} from "react";

import {
  contractABI,
  contractAddress,
  marketplaceAddress,
} from "../utils/constants";
import { cards } from "../utils/cards";
import { getEthereumContract } from "../utils/connect";
import { WebContext } from "./WebContext";

export const TradingCardMinterContext = createContext();

const { ethereum } = window;

export function TradingCardMinterProvider({ children }) {
  const [currentAccount, setCurrentAccount] = useState(null);
  const { setShowAlert, setSuccess, setAlertMessage } = useContext(WebContext);
  const player1Ref = useRef();
  const player2Ref = useRef();

  const mintTradingCard = async () => {
    try {
      if (ethereum) {
        const contract = getEthereumContract(contractAddress, contractABI);

        const transaction = await contract.requestNewCard();
        await transaction.wait();
        console.log(
          `1 Card successfully sent - Transaction hash: ${transaction.hash}`
        );
      } else {
        console.log("Ethereum is not present");
      }
    } catch (err) {
      console.log(err);
    }
  };

  const getCardsUnderAddress = async () => {
    try {
      if (ethereum) {
        if (!currentAccount) {
          return [];
        }
        const contract = getEthereumContract(contractAddress, contractABI);

        const data = await contract.getCardsUnderOwner(currentAccount);
        const ids = data.map((card) => ({
          card: cards[card.cardId.toNumber()],
          tokenId: card.tokenId.toNumber(),
        }));
        return ids;
      }
    } catch (err) {
      console.log(err);
    }
    return [];
  };

  const approveMarketplaceContract = async (tokenId) => {
    try {
      if (ethereum) {
        const contract = getEthereumContract(contractAddress, contractABI);

        const transaction = await contract.approve(marketplaceAddress, tokenId);
        await transaction.wait();
        console.log(
          `Approved marketplace contract to sell card with id ${tokenId} - Transaction hash: ${transaction.hash}`
        );
      } else {
        console.log("Ethereum is not present");
      }
    } catch (err) {
      console.log(err);
    }
  };

  const requestNewPack = async () => {
    try {
      if (ethereum) {
        const contract = getEthereumContract(contractAddress, contractABI);

        const transaction = await contract.buyCardPack();
        await transaction.wait();
        console.log(
          `1 Pack successfully sent - Transaction hash: ${transaction.hash}`
        );
      } else {
        console.log("Ethereum is not present");
      }
    } catch (err) {
      console.log(err);
    }
  };

  useEffect(() => {
    const checkIfWalletIsConnected = async () => {
      try {
        if (!ethereum) {
          setAlertMessage("Make sure you have metamask!");
          setShowAlert(true);
          setSuccess(false);
          return;
        }

        const accounts = await ethereum.request({ method: "eth_accounts" });

        if (accounts.length) {
          const account = accounts[0];
          console.log("Found an authorized account: ", account);
          setCurrentAccount(account);
        } else {
          console.log("No authorized account found");
        }
      } catch (err) {
        console.log(err);
      }
    };
    checkIfWalletIsConnected();
  }, [setAlertMessage, setShowAlert, setSuccess]);

  return (
    <TradingCardMinterContext.Provider
      value={{
        mintTradingCard,
        getCardsUnderAddress,
        currentAccount,
        player1Ref,
        player2Ref,
        approveMarketplaceContract,
        requestNewPack,
      }}
    >
      {children}
    </TradingCardMinterContext.Provider>
  );
}
