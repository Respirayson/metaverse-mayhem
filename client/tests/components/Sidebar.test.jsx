import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Sidebar from "../../src/components/SideBar";
import { NftMarketplaceContext } from "../../src/context/NftMarketplace";
import { BrowserRouter } from "react-router-dom";

describe("Sidebar", () => {
  it("renders the component with correct content", () => {
    const mockGetProceeds = vi.fn().mockResolvedValue("0.0");
    const mockWithdrawProceeds = vi.fn().mockResolvedValue(undefined);

    const { container } = render(
      <BrowserRouter><Sidebar url="/dashboard" /></BrowserRouter>,
      {
        wrapper: ({ children }) => (
          <NftMarketplaceContext.Provider
            value={{
              getProceeds: mockGetProceeds,
              withdrawProceeds: mockWithdrawProceeds,
            }}
          >
            {children}
          </NftMarketplaceContext.Provider>
        ),
      }
    );

    const withdrawButton = document.querySelector('#withdraw');

    expect(mockGetProceeds).toHaveBeenCalledTimes(1);
    expect(mockWithdrawProceeds).toHaveBeenCalledTimes(0);

    fireEvent.click(withdrawButton);

    expect(mockWithdrawProceeds).toHaveBeenCalledTimes(1);

    expect(container).toMatchSnapshot();
  });
});
