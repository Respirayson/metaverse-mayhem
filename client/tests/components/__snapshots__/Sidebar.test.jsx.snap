// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sidebar > renders the component with correct content 1`] = `
<div>
  <div
    class="sm:flex hidden mr-10 relative z-[100000]"
  >
    <div
      class="flex justify-between items-center flex-col sticky top-5 h-[80vh]"
    >
      <div
        class="flex-1 flex flex-col justify-between items-center bg-[#1c1c24] rounded-[20px] w-[76px] py-4 mt-12"
      >
        <div
          class="flex flex-col justify-center items-center gap-3 relative"
        >
          <div
            class="w-[48px] h-[48px] rounded-[10px] false flex justify-center items-center cursor-pointer"
            id="all-listings"
            style="opacity: 0;"
          >
            <img
              alt="logo"
              class="w-1/2 h-1/2 grayscale"
              src="/marketplace.svg"
            />
          </div>
          <div
            class="w-[48px] h-[48px] rounded-[10px] false flex justify-center items-center cursor-pointer"
            id="my-listings"
            style="opacity: 0;"
          >
            <img
              alt="logo"
              class="w-1/2 h-1/2 grayscale"
              src="/my-listings.svg"
            />
          </div>
          <div
            class="w-[48px] h-[48px] rounded-[10px] false flex justify-center items-center cursor-pointer"
            id="store"
            style="opacity: 0;"
          >
            <img
              alt="logo"
              class="w-1/2 h-1/2 grayscale"
              src="/store.svg"
            />
          </div>
          <div
            class="w-[48px] h-[48px] rounded-[10px] false flex justify-center items-center cursor-pointer"
            id="create-listing"
            style="opacity: 0;"
          >
            <img
              alt="logo"
              class="w-1/2 h-1/2 grayscale"
              src="/create-listing.svg"
            />
          </div>
        </div>
        <div
          class="w-[48px] h-[48px] rounded-[10px] undefined flex justify-center items-center false"
          id="withdraw"
          style="opacity: 0;"
        >
          <img
            alt="logo"
            class="w-1/2 h-1/2"
            src="/wallet.svg"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
